### **M7T1**
Crea 4 variables, a una assigna-li la teva edat, a un altre el teu nom, a un altre el teu cognom i a un altre el valor true si estàs a l'atur o false si no estàs.

Imprimeix aquestes 4 variables anteriors juntament amb la frase "El meu nom és: X, el meu cognom és: Y i tinc Z anys. Estic a l'atur? W"" on X, Y, Z i W són els valors de les variables que has creat abans.

Demana 3 dades més mitjançant 3 promps i printa una frase amb els 3 valors que has demanat (pots inventar la temàtica de les dades que demanis).

In [ ]:
X='Noelia'
Y='Solano'
Z=24
W=True
print(f"El meu nom és: {X}, el meu cognom és: {Y} i tinc {Z} anys. Estic a l'atur? {W}")
A = input(f"T'agrada llegir {X}? Marca sí o no, si us plau: ")
B = input("Quin és l'últim llibre que has llegit? ")
C = input("Qui n'és l'autor/a? ")
print(f"T'agrada llegir? {A}. L'últim llibre que recordo haver llegit va ser {B} de {C}.")

<class 'bool'>
El meu nom és: Noelia, el meu cognom és: Solano i tinc 24 anys. Estic a l'atur? True
T'agrada llegir? si. L'últim llibre que recordo haver llegit va ser el alquimista de paulo coelho.


### **M7T2**
Crea un programa que demani l'edat a l'usuari/ària. Si aquesta és igual o major a 18, el programa ha d'imprimir per pantalla "Ets major d'edat" si no "NO ets major d'edat."

In [4]:
edat=int(input('Quants anys tens?'))
if edat>18:
    print("Ets major d'edat")
else:
    print("No ets major d'edat")

No ets major d'edat


### **M7T3**
Modifica el programa anterior i fes que la resposta segons l'edat segueixi aquest algoritme:

Si l'edat és negativa, el programa ha d'imprimir: "una edat negativa no és possible per a una persona."

Si l'edat està entre 0 i 17 (ambdós inclosos), el programa ha d'imprimir: "ets MENOR d'edat."

Si l'edat està entre 18 i 125 (ambdós inclosos), el programa ha d'imprimir: "ets MAJOR d'edat."

Si l'edat és major que 125, el programa ha d'imprimir: "una edat tan alta és difícil de veure en una persona."

In [14]:
edat=int(input('Quants anys tens?'))
if edat<=0:
    print("Una edat negativa no és possible per a una persona")
elif edat<=17:
    print("ets MENOR d'edat")
elif edat<=125:
    print("Ets MAJOR d'edat")
else:
    print("Una edat tan alta és difícil de veure en una persona")

Ets MAJOR d'edat


*NOTA:* Siempre que utilizemos return, hay que guardar el resultado a la hora de llmar la funcion en una variable para luego poder imprimir este resultado. Y si ademas queremos hacer algo con el resultado hay que guardarlo en una variable si o si

### **M7T4**
Crea una funció que demani el nom a l'usuari i mostri un missatge de salutació.

In [ ]:
def salutacio():
    nom=input('Com et dius?')
    print(f'Benvolgut/da {nom}, prem qualsevol tecla per continuar.')

salutacio()

Benvolgut/da Noelia, prem qualsevol tecla per continuar


### **M7T5**
Crea una funció que rebi un preu i retorni el preu amb IVA (21% per defecte).

A continuació, demana un preu a l’usuari i mostra el resultat amb l’IVA aplicat.

In [18]:
def compte(preu):
    total=preu*1.21
    return total

preu=float(input('Indiqueu el preu a continuació:'))
c=compte(preu)
print(f'El total del compte és {c:.2f}€, IVA inclòs.') #:.nf --> redondea al numero de decimales n que se muestra

El total del compte és 121.00€, IVA inclòs.


### **M7T6**
Crea una funció que rebi un número enter com a argument i retorni True si el número és parell, o False si és senar.

A continuació, demana un número a l'usuari, utilitza la funció i mostra un missatge indicant si el número és parell o senar.

In [1]:
def proba(num):
    return num % 2 == 0

num=int(input('Indica un número enter:'))
TF=proba(num)
if TF:
    print('El número es parell')
else:
    print('El número es senar')

El número es senar


### **M7T7**

**Exercici 1**

Crea una llista amb tres tasques inicials: "netejar la cuina", "treure el gos a passejar" i "portar a la nena al col·legi". Un cop creada, imprimeix-la.

Després afegeix al final de la llista, les tasques: "fer esport", "meditar". Torna a imprimir-la per veure que s'han afegit.

Elimina la tasca "netejar la cuina" i torna a imprimir la llista.

Finalment, afegeix la tasca "cuinar" a la segona posició de la llista.

**Exercici 2**

Crea un diccionari on puguem determinar quin dia es fa cada tasca. Assigna un dia diferent de la setmana a totes les tasques anteriors.

Printa la tasca del dilluns.

In [24]:
#1
tasques=["netejar la cuina", "treure el gos a passejar", "portar a la nena al col·legi"]
tasques.extend(["fer esport", "meditar"]) #tareas += ["fer esport", "meditar"]
print(tasques)
tasques.pop(0)
print(tasques)
tasques.insert(1, "cuinar")
print(tasques)
#2
dias=["dilluns", "dimarts", "dimecres", "dijous", "divendres"] 
calendari=dict(zip(dias,tasques)) #zip: unimos ambas listas en un diccionario por iteración paralela
print(calendari)
print(f'La tasca del dilluns és: {calendari["dilluns"]}') #print(dict_tareas.get("lunes"))-->para diccionarios únicamente

['netejar la cuina', 'treure el gos a passejar', 'portar a la nena al col·legi', 'fer esport', 'meditar']
['treure el gos a passejar', 'portar a la nena al col·legi', 'fer esport', 'meditar']
['treure el gos a passejar', 'cuinar', 'portar a la nena al col·legi', 'fer esport', 'meditar']
{'dilluns': 'treure el gos a passejar', 'dimarts': 'cuinar', 'dimecres': 'portar a la nena al col·legi', 'dijous': 'fer esport', 'divendres': 'meditar'}
La tasca del dilluns és: treure el gos a passejar


In [26]:
calendari['dissabte']='fer la bugada' #ejemplo de añadir clave y valor en diccionario
print(calendari)

{'dilluns': 'treure el gos a passejar', 'dimarts': 'cuinar', 'dimecres': 'portar a la nena al col·legi', 'dijous': 'fer la bugada', 'divendres': 'fer esport', 'dissabte': 'fer la bugada'}


### **M7T8**
Crea una funció que faci el següent:

Creï una llista buida.
Demani a l’usuari 3 paraules, una per una.
Si la paraula comença per 'A' o 'a' i té menys de 7 lletres, afegeix-la a la llista.
Si no compleix aquestes condicions, no facis res.
Un cop demanades les 3 paraules i avaluades, imprimeix la llista completa.

Finalment, crida la funció perquè s’executi.

In [60]:
def test():
    lista=[]
    for paraula in range(3):
        paraula=input('Indiqui una paraula:')
        if paraula.startswith(('A','a')) and len(paraula)<7:
            lista.append(paraula)
    return lista
t=test()
print(f'La llista vàlida és: {t}')


La llista vàlida és: ['ala', 'ana']


*NOTA:* la diferencia entre **for y while** es que para establecer rango() en while debemos utilizar un contador manual 

*NOTA:* al recorrer un diccionario, si usamos for nos dara solo la clave y no valores a no ser que hagamos:

In [66]:
diccionario = {10:'diez', 20:'veinte', 30:'treinta'}
for clave in diccionario:
    print(clave, ':', diccionario[clave])

10 : diez
20 : veinte
30 : treinta


### **M7T9**
Crea una funció que rebi com a paràmetre el nombre de paraules que l’usuari/ària vol introduir. A continuació:

Demana tantes paraules com hagi indicat l’usuari/ària i desa-les en una llista.

Mostra per pantalla totes les paraules introduïdes.

Finalment, abans de cridar la funció, demana a l’usuari/ària quantes paraules vol introduir.

In [ ]:
def funcio(num):
    llista=[]
    for i in range(num):
        paraules=input(f'Indiqui la paraula número {i+1}: ')
        llista.append(paraules)
    print(f'Aquesta és la llista resultant: {llista}.') #Estamos usando un print no un return, el f'' se debe poner en este print no fuera de la función para que se muestre bien.
num=int(input('Indiqui el número de paraules que vol introduir: '))
funcio(num)


Aquesta és la llista resultant: ['s', 't']


### **M7T10**
Defineix una funció que rebi una llista de números i retorni el sumatori de tots els elements (sense utilitzar sum(), fes el càlcul manualment amb un bucle).

Crea una llista amb 10 números.
Calcula la mitjana fent servir la funció del sumatori.
Mostra per pantalla el sumatori i la mitjana.

In [147]:
def sumatori(llista):
    càlcul=0
    for i in llista: #funciona como un indicador de recorrido-->i: para cada elemto en la lista
        càlcul+=i #aqui entiende que tiene que sumer el elemento de forma acumulativa
    return càlcul
llista=[0,1,2,3,4,5,6,7,8,9]
s=sumatori(llista)
m=s/len(llista)
print(f'El sumatori resultant de {llista} es {sumatori(llista)} y la mitjana es {m}.')

El sumatori resultant de [0, 1, 2, 3, 4, 5, 6, 7, 8, 9] es 45 y la mitjana es 4.5.


In [ ]:
def sumatori():# si no damos el argumento aqui no hay que usarlo al guardar tampoco --> s=sumatori()
    càlcul=0
    for i in llista: #funciona como un indicador de recorrido-->i: para cada elemto en la lista
        càlcul+=i #aqui entiende que tiene que sumer el elemento de forma acumulativa
    return càlcul
llista=[0,1,2,3,4,5,6,7,8,9]
s=sumatori()
m=s/len(llista)
print(f'El sumatori resultant de {llista} es {sumatori()} y la mitjana es {m}.')

El sumatori resultant de [0, 1, 2, 3, 4, 5, 6, 7, 8, 9] es 45 y la mitjana es 4.5.


### **M7T11**
Crea una funció que faci el següent:

Demani a l’usuari/ària 5 articles i el seu preu corresponent.

Desi aquests articles i preus en un diccionari, on la clau sigui el nom de l’article i el valor, el seu preu.

Retorni el diccionari amb les dades introduïdes.

Un cop creat el diccionari, imprimeix només els articles que costin menys de 10 euros.

In [ ]:
def cistella():
    diccionari={}
    for i in range(5):
        article=input(f'Introdueix un article {i+1}/5: ')
        preu=int(input('Introdueix un preu: '))
        diccionari[article]=preu
    return diccionari

c=cistella()
print(c)

articles_barats=[]

for article, preu in c.items():
     if preu<10:
         articles_barats.append(article)
print(f'Els articles menors de de 10 euros són: {articles_barats}')

{'g': 55, 'e': 3, 'u': 77, 't': 2, 's': 32}
Els articles menors de de 10 euros són: ['e', 't']


*NOTA:* Cuando Python ve una variable dentro de una función (por ejemplo c dentro de cistella_barata()), no la busca solo en un sitio.
Sigue este orden (de adentro hacia afuera):

L – Local: dentro de la función actual.

E – Enclosing: dentro de funciones que envuelven a la actual (en caso de funciones anidadas).

G – Global: en el archivo/script donde se ejecuta el programa.

B – Built-in: en los nombres integrados de Python (como len, print, etc.).

Por esto si nuestra segunda función cistella_barata() estuviera primero no funcionaria porque no encontraria c=cistella() con LEGB.

In [ ]:
#Solución alternativa con otra función
def cistella():
    diccionari={}
    for i in range(5):
        article=input(f'Introdueix un article {i+1}/5: ')
        preu=int(input('Introdueix un preu: '))
        diccionari[article]=preu
    return diccionari

c=cistella() #guardamos la función en ambito global
print(c)

def cistella_barata(menys10): #Función reutilizable: para que funcione mejor el codigo es preferible añadir una variable no externa, así python entiende que depende por el momento no es dependiente de algo externo, no busca con LEGB y es reutilizable hasta que la llamemos (o guardemos) con una variable externa por argumento.
    articles_barats=[]
    for article, preu in menys10.items():
       if preu<10:
        articles_barats.append(article)
    print(f'Els articles menors de 10 euros són: {articles_barats}')

cistella_barata(c) #--> aqui referenciamos a la función c

{'y': 5, 'r': 44, 'e': 3, 'w': 22, 't': 1}
Els articles menors de 10 euros són: ['y', 'e', 't']


### **M7T12**
En aquest exercici, treballaràs amb el dataset Salut_i_son.csv utilitzant la biblioteca Pandas de Python. Realitzaràs diverses tasques d’exploració, filtres, agrupacions i modificació de columnes.

In [178]:
import pandas as pd
df=pd.read_csv('csvs/Salut_i_son.csv')
df

Person ID  Gender  Age            Occupation  Sleep Duration  \
0            1    Male   27     Software Engineer             6.1   
1            2    Male   28                Doctor             6.2   
2            3    Male   28                Doctor             6.2   
3            4    Male   28  Sales Representative             5.9   
4            5    Male   28  Sales Representative             5.9   
..         ...     ...  ...                   ...             ...   
369        370  Female   59                 Nurse             8.1   
370        371  Female   59                 Nurse             8.0   
371        372  Female   59                 Nurse             8.1   
372        373  Female   59                 Nurse             8.1   
373        374  Female   59                 Nurse             8.1   

     Quality of Sleep  Physical Activity Level  Stress Level BMI Category  \
0                   6                       42             6   Overweight   
1                   6                       60             8       Normal   
2                   6                       60             8       Normal   
3                   4                       30             8        Obese   
4                   4                       30             8        Obese   
..                ...                      ...           ...          ...   
369                 9                       75             3   Overweight   
370                 9                       75             3   Overweight   
371                 9                       75             3   Overweight   
372                 9                       75             3   Overweight   
373                 9                       75             3   Overweight   

    Blood Pressure  Heart Rate  Daily Steps Sleep Disorder  
0           126/83          77         4200            NaN  
1           125/80          75        10000            NaN  
2           125/80          75        10000            NaN  
3           140/90          85         3000    Sleep Apnea  
4           140/90          85         3000    Sleep Apnea  
..             ...         ...          ...            ...  
369         140/95          68         7000    Sleep Apnea  
370         140/95          68         7000    Sleep Apnea  
371         140/95          68         7000    Sleep Apnea  
372         140/95          68         7000    Sleep Apnea  
373         140/95          68         7000    Sleep Apnea  

[374 rows x 13 columns]

1. Exploració de dades
Carrega el dataset Salut_i_son.csv i respon les següents preguntes:

Quantes columnes i quants registres té el dataset?
Quantes columnes són de tipus enter? I quantes de tipus decimal?
Quina és l’edat màxima registrada?
Quina és la duració mínima de la son?
Quina és la mitjana de la qualitat de la son?
Explica com has trobat cadascuna de les respostes

In [179]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 374 entries, 0 to 373
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Person ID                374 non-null    int64  
 1   Gender                   374 non-null    object 
 2   Age                      374 non-null    int64  
 3   Occupation               374 non-null    object 
 4   Sleep Duration           374 non-null    float64
 5   Quality of Sleep         374 non-null    int64  
 6   Physical Activity Level  374 non-null    int64  
 7   Stress Level             374 non-null    int64  
 8   BMI Category             374 non-null    object 
 9   Blood Pressure           374 non-null    object 
 10  Heart Rate               374 non-null    int64  
 11  Daily Steps              374 non-null    int64  
 12  Sleep Disorder           155 non-null    object 
dtypes: float64(1), int64(7), object(5)
memory usage: 38.1+ KB


In [180]:
df.describe()

Person ID         Age  Sleep Duration  Quality of Sleep  \
count  374.000000  374.000000      374.000000        374.000000   
mean   187.500000   42.184492        7.132086          7.312834   
std    108.108742    8.673133        0.795657          1.196956   
min      1.000000   27.000000        5.800000          4.000000   
25%     94.250000   35.250000        6.400000          6.000000   
50%    187.500000   43.000000        7.200000          7.000000   
75%    280.750000   50.000000        7.800000          8.000000   
max    374.000000   59.000000        8.500000          9.000000   

       Physical Activity Level  Stress Level  Heart Rate   Daily Steps  
count               374.000000    374.000000  374.000000    374.000000  
mean                 59.171123      5.385027   70.165775   6816.844920  
std                  20.830804      1.774526    4.135676   1617.915679  
min                  30.000000      3.000000   65.000000   3000.000000  
25%                  45.000000      4.000000   68.000000   5600.000000  
50%                  60.000000      5.000000   70.000000   7000.000000  
75%                  75.000000      7.000000   72.000000   8000.000000  
max                  90.000000      8.000000   86.000000  10000.000000

1. **RESPUESTA:** El dataframe contiene 374 registros y 13 columnas. Hay una columna de tipo decimal y 7 de tipo entero. La edad maxima registrada es 59 y el sueño minimo es 5,8. La media de la calidad de sueño es 7,13. Registros, columnas y tipo de variable los encontré con df.info() y los demás con df.describe().

2. Filtres

Classifica les dades en dos grups de persones segons els criteris següents:

Grup 1: Persones de més de 40 anys, amb una qualitat de la son superior a 8 i que fan més de 5000 passes al dia.

In [202]:
#Grup 1:
df[(df['Age']>40)
   & (df['Quality of Sleep']>8)
   & (df['Daily Steps']>5000)]

Person ID  Gender  Age Occupation  Sleep Duration  Quality of Sleep  \
341        342  Female   56     Doctor             8.2                 9   
342        343  Female   56     Doctor             8.2                 9   
343        344  Female   57      Nurse             8.1                 9   
344        345  Female   57      Nurse             8.2                 9   
345        346  Female   57      Nurse             8.2                 9   
346        347  Female   57      Nurse             8.2                 9   
347        348  Female   57      Nurse             8.2                 9   
348        349  Female   57      Nurse             8.2                 9   
349        350  Female   57      Nurse             8.1                 9   
350        351  Female   57      Nurse             8.1                 9   
351        352  Female   57      Nurse             8.1                 9   
352        353  Female   58      Nurse             8.0                 9   
353        354  Female   58      Nurse             8.0                 9   
354        355  Female   58      Nurse             8.0                 9   
355        356  Female   58      Nurse             8.0                 9   
356        357  Female   58      Nurse             8.0                 9   
357        358  Female   58      Nurse             8.0                 9   
358        359  Female   59      Nurse             8.0                 9   
359        360  Female   59      Nurse             8.1                 9   
360        361  Female   59      Nurse             8.2                 9   
361        362  Female   59      Nurse             8.2                 9   
362        363  Female   59      Nurse             8.2                 9   
363        364  Female   59      Nurse             8.2                 9   
364        365  Female   59      Nurse             8.0                 9   
365        366  Female   59      Nurse             8.0                 9   
366        367  Female   59      Nurse             8.1                 9   
367        368  Female   59      Nurse             8.0                 9   
368        369  Female   59      Nurse             8.1                 9   
369        370  Female   59      Nurse             8.1                 9   
370        371  Female   59      Nurse             8.0                 9   
371        372  Female   59      Nurse             8.1                 9   
372        373  Female   59      Nurse             8.1                 9   
373        374  Female   59      Nurse             8.1                 9   

     Physical Activity Level  Stress Level   BMI Category Blood Pressure  \
341                       90             3  Normal Weight         118/75   
342                       90             3  Normal Weight         118/75   
343                       75             3     Overweight         140/95   
344                       75             3     Overweight         140/95   
345                       75             3     Overweight         140/95   
346                       75             3     Overweight         140/95   
347                       75             3     Overweight         140/95   
348                       75             3     Overweight         140/95   
349                       75             3     Overweight         140/95   
350                       75             3     Overweight         140/95   
351                       75             3     Overweight         140/95   
352                       75             3     Overweight         140/95   
353                       75             3     Overweight         140/95   
354                       75             3     Overweight         140/95   
355                       75             3     Overweight         140/95   
356                       75             3     Overweight         140/95   
357                       75             3     Overweight         140/95   
358                       75             3     Overw

Grup 2: Persones que treballen com a enginyers/eres o enginyers/eres de software i tenen una heart rate superior a 60.

In [201]:
#Grup 2:
df[((df['Occupation']=='Engineer') | (df['Occupation']=='Software Engineer'))
   & (df['Heart Rate']>60)]



Person ID  Gender  Age         Occupation  Sleep Duration  \
0            1    Male   27  Software Engineer             6.1   
5            6    Male   28  Software Engineer             5.9   
50          51    Male   32           Engineer             7.5   
51          52    Male   32           Engineer             7.5   
84          85    Male   35  Software Engineer             7.5   
..         ...     ...  ...                ...             ...   
334        335  Female   54           Engineer             8.4   
335        336  Female   54           Engineer             8.4   
336        337  Female   54           Engineer             8.4   
337        338  Female   54           Engineer             8.4   
338        339  Female   54           Engineer             8.5   

     Quality of Sleep  Physical Activity Level  Stress Level   BMI Category  \
0                   6                       42             6     Overweight   
5                   4                       30             8          Obese   
50                  8                       45             3         Normal   
51                  8                       45             3         Normal   
84                  8                       60             5  Normal Weight   
..                ...                      ...           ...            ...   
334                 9                       30             3         Normal   
335                 9                       30             3         Normal   
336                 9                       30             3         Normal   
337                 9                       30             3         Normal   
338                 9                       30             3         Normal   

    Blood Pressure  Heart Rate  Daily Steps Sleep Disorder  
0           126/83          77         4200            NaN  
5           140/90          85         3000       Insomnia  
50          120/80          70         8000            NaN  
51          120/80          70         8000            NaN  
84          120/80          70         8000            NaN  
..             ...         ...          ...            ...  
334         125/80          65         5000            NaN  
335         125/80          65         5000            NaN  
336         125/80          65         5000            NaN  
337         125/80          65         5000            NaN  
338         125/80          65         5000            NaN  

[67 rows x 13 columns]

3. Agrupacions
Agrupa les dades del dataset original per calcular valors mitjans:

Calcula la mitjana de la duració de la son agrupada per gènere.

In [203]:
df.groupby('Gender')[['Sleep Duration']].mean()

Sleep Duration
Gender                
Female        7.229730
Male          7.036508

Calcula la mitjana de la duració de la son agrupada per gènere i ocupació.

In [205]:
df.groupby(['Gender', 'Occupation'])[['Sleep Duration']].mean()

Sleep Duration
Gender Occupation                          
Female Accountant                  7.111111
       Doctor                      8.200000
       Engineer                    8.425000
       Lawyer                      7.150000
       Manager                     6.900000
       Nurse                       7.063014
       Scientist                   6.000000
       Teacher                     6.705714
Male   Accountant                  7.200000
       Doctor                      6.934783
       Engineer                    7.535484
       Lawyer                      7.422222
       Sales Representative        5.900000
       Salesperson                 6.403125
       Software Engineer           6.750000
       Teacher                     6.580000

4. Modificació de columnes
Afegeix una nova columna al DataFrame, anomenada "Poca Son", amb el valor True si la persona dorm menys de 7 hores, i False en cas contrari.

In [212]:
df['Little Sleep']=df['Sleep Duration']<7
df

Person ID  Gender  Age            Occupation  Sleep Duration  \
0            1    Male   27     Software Engineer             6.1   
1            2    Male   28                Doctor             6.2   
2            3    Male   28                Doctor             6.2   
3            4    Male   28  Sales Representative             5.9   
4            5    Male   28  Sales Representative             5.9   
..         ...     ...  ...                   ...             ...   
369        370  Female   59                 Nurse             8.1   
370        371  Female   59                 Nurse             8.0   
371        372  Female   59                 Nurse             8.1   
372        373  Female   59                 Nurse             8.1   
373        374  Female   59                 Nurse             8.1   

     Quality of Sleep  Physical Activity Level  Stress Level BMI Category  \
0                   6                       42             6   Overweight   
1                   6                       60             8       Normal   
2                   6                       60             8       Normal   
3                   4                       30             8        Obese   
4                   4                       30             8        Obese   
..                ...                      ...           ...          ...   
369                 9                       75             3   Overweight   
370                 9                       75             3   Overweight   
371                 9                       75             3   Overweight   
372                 9                       75             3   Overweight   
373                 9                       75             3   Overweight   

    Blood Pressure  Heart Rate  Daily Steps Sleep Disorder  Little Sleep  
0           126/83          77         4200            NaN          True  
1           125/80          75        10000            NaN          True  
2           125/80          75        10000            NaN          True  
3           140/90          85         3000    Sleep Apnea          True  
4           140/90          85         3000    Sleep Apnea          True  
..             ...         ...          ...            ...           ...  
369         140/95          68         7000    Sleep Apnea         False  
370         140/95          68         7000    Sleep Apnea         False  
371         140/95          68         7000    Sleep Apnea         False  
372         140/95          68         7000    Sleep Apnea         False  
373         140/95          68         7000    Sleep Apnea         False  

[374 rows x 14 columns]

Elimina les últimes 4 columnes del DataFrame original i desa el nou DataFrame en una variable anomenada sleep_info.

In [218]:
#Opción 1
sleep_info=df.drop(['Heart Rate', 'Daily Steps', 'Sleep Disorder', 'Little Sleep'], axis=1)
sleep_info

Person ID  Gender  Age            Occupation  Sleep Duration  \
0            1    Male   27     Software Engineer             6.1   
1            2    Male   28                Doctor             6.2   
2            3    Male   28                Doctor             6.2   
3            4    Male   28  Sales Representative             5.9   
4            5    Male   28  Sales Representative             5.9   
..         ...     ...  ...                   ...             ...   
369        370  Female   59                 Nurse             8.1   
370        371  Female   59                 Nurse             8.0   
371        372  Female   59                 Nurse             8.1   
372        373  Female   59                 Nurse             8.1   
373        374  Female   59                 Nurse             8.1   

     Quality of Sleep  Physical Activity Level  Stress Level BMI Category  \
0                   6                       42             6   Overweight   
1                   6                       60             8       Normal   
2                   6                       60             8       Normal   
3                   4                       30             8        Obese   
4                   4                       30             8        Obese   
..                ...                      ...           ...          ...   
369                 9                       75             3   Overweight   
370                 9                       75             3   Overweight   
371                 9                       75             3   Overweight   
372                 9                       75             3   Overweight   
373                 9                       75             3   Overweight   

    Blood Pressure  
0           126/83  
1           125/80  
2           125/80  
3           140/90  
4           140/90  
..             ...  
369         140/95  
370         140/95  
371         140/95  
372         140/95  
373         140/95  

[374 rows x 10 columns]

In [ ]:
#Opción 2
sleep_info=df.iloc[:, :-4]
sleep_info

Person ID  Gender  Age            Occupation  Sleep Duration  \
0            1    Male   27     Software Engineer             6.1   
1            2    Male   28                Doctor             6.2   
2            3    Male   28                Doctor             6.2   
3            4    Male   28  Sales Representative             5.9   
4            5    Male   28  Sales Representative             5.9   
..         ...     ...  ...                   ...             ...   
369        370  Female   59                 Nurse             8.1   
370        371  Female   59                 Nurse             8.0   
371        372  Female   59                 Nurse             8.1   
372        373  Female   59                 Nurse             8.1   
373        374  Female   59                 Nurse             8.1   

     Quality of Sleep  Physical Activity Level  Stress Level BMI Category  \
0                   6                       42             6   Overweight   
1                   6                       60             8       Normal   
2                   6                       60             8       Normal   
3                   4                       30             8        Obese   
4                   4                       30             8        Obese   
..                ...                      ...           ...          ...   
369                 9                       75             3   Overweight   
370                 9                       75             3   Overweight   
371                 9                       75             3   Overweight   
372                 9                       75             3   Overweight   
373                 9                       75             3   Overweight   

    Blood Pressure  
0           126/83  
1           125/80  
2           125/80  
3           140/90  
4           140/90  
..             ...  
369         140/95  
370         140/95  
371         140/95  
372         140/95  
373         140/95  

[374 rows x 10 columns]

### **M7T13**
La nostra empresa es dedica a realitzar anàlisis de dades geogràfiques i socials, i un client ens ha sol·licitat una eina per identificar dades bàsiques de cada continent de manera automàtica.

Per això, cal crear una funció que rebi dos paràmetres:

Un DataFrame amb informació de països (com el del fitxer Country.xlsx).
El nom d’un continent (com a cadena de text).

La funció ha de retornar un diccionari amb les següents dades relacionades amb el continent indicat:

Nombre de regions diferents que hi ha dins del continent.
Nombre de països dins del continent.
Superfície màxima (SurfaceArea) registrada al continent.
Superfície mínima registrada.
Població mitjana (Population) dels països del continent.
Mitjana de l’esperança de vida (LifeExpectancy) dels països del continent.

Finalment, carrega el fitxer adjunt amb la següent instrucció:

data = pd.read_csv("arxiu csv", sep=";", encoding="latin-1")

Executa la funció per a diversos continents i mostra els resultats obtinguts.

In [312]:
data = pd.read_csv("csvs/Country(in).csv", sep=";", encoding="latin-1")
# Como utilizar el filtrado con funcion
def funcio(data, continent):
    # Creamos un diccionario contenedor, queremos que solo tenga en cuenta lo referente a continentes, con filtrado seleccionamos solo continentes y guardamos en un nuevo dataframe
    data2 = data[data['Continent'] == continent]
    diccionari = {
        'Numero de regions' : data2['Region'].nunique(),
        'Numero de paisos' : data2['Name'].nunique(),
        'Superficie maxima' : data2['SurfaceArea'].max(),
        'Superficie minima' : data2['SurfaceArea'].min(),
        'Poblacio minima' : data2['Population'].mean(),
        'Esperança de vida mitja' : data2['LifeExpectancy'].mean()
    }
    return diccionari
funcio(data, 'Asia')

{'Numero de regions': 4,
 'Numero de paisos': 51,
 'Superficie maxima': np.float64(9572900.0),
 'Superficie minima': np.float64(18.0),
 'Poblacio minima': np.float64(72647562.74509804),
 'Esperança de vida mitja': np.float64(67.44117647058823)}

In [315]:
data['Continent'].unique() #Podemos ver todos los continentes unicos que hay

array(['North America', 'Asia', 'Africa', 'Europe', 'South America',
       'Oceania', 'Antarctica'], dtype=object)

In [ ]:
#Apuntes con groupby
data.groupby('Region').ngroups #numero de regiones
data.groupby('Name').groups #ver paises y sus indices --> .groups nos devuelve un diccionario
data.groupby('Continent').groups.keys() #ver solo nombres de las continentes
data.groupby('Continent').get_group('Asia') #si ponemos algo dentro de .groups.keys(), buscara expecificamente ese valor
#Sin groupby, solo o despues de filtrado.
data['Region'].nunique() #numero de regiones
data['Name'].count() #numero de paises
#Filtrado
data2 = df[df['Continent'] == c]
data2['Name'].count()

{'Antarctica': [11, 12, 34, 93, 187], 'Australia and New Zealand': [14, 38, 52, 154, 162], 'Baltic Countries': [67, 126, 128], 'British Islands': [76, 101], 'Caribbean': [0, 3, 6, 13, 24, 31, 51, 53, 58, 60, 81, 86, 96, 107, 115, 122, 145, 146, 172, 204, 213, 226, 228, 229], 'Central Africa': [2, 36, 43, 44, 45, 75, 84, 196, 205], 'Central America': [27, 50, 88, 94, 135, 156, 165, 192], 'Eastern Africa': [17, 48, 57, 64, 68, 100, 111, 133, 143, 147, 148, 150, 179, 182, 194, 202, 218, 219, 237, 238], 'Eastern Asia': [41, 92, 109, 116, 129, 141, 173, 217], 'Eastern Europe': [22, 26, 55, 97, 132, 171, 180, 181, 198, 220], 'Melanesia': [70, 152, 170, 190, 231], 'Micronesia': [74, 90, 114, 136, 142, 161, 169], 'Micronesia/Caribbean': [221], 'Middle East': [7, 9, 16, 23, 54, 77, 103, 105, 108, 117, 119, 163, 176, 178, 183, 203, 215, 234], 'Nordic Countries': [59, 69, 73, 104, 159, 189, 200], 'North America': [28, 37, 87, 195, 223], 'Northern Africa': [61, 63, 65, 121, 130, 184, 214], 'Polyne

*APUNTES:*

Usa filtrado (df[df['Columna'] == valor]) cuando:

Solo te interesa un grupo concreto (por ejemplo, un continente, una región, un país).
Quieres calcular estadísticas para ese grupo específico.
🧠 Ejemplo:
“Quiero la media de población solo de Europa.”
👉 Usa filtrado.

Usa groupby() cuando:
Quieres comparar varios grupos a la vez.
Necesitas calcular valores para todos los grupos en una sola operación.
🧠 Ejemplo:
“Quiero la media de población de todos los continentes.”
👉 Usa groupby('Continent').

🔸 Filtrado → un grupo.

🔸 Groupby → varios grupos.

### **M7T14**
La nostra empresa està col·laborant amb una ONG que estudia la distribució de l'espai i dels recursos geogràfics en relació amb la població mundial. L’objectiu és identificar quins països tenen menys espai disponible per habitant, fet que pot estar relacionat amb situacions de sobrepoblació, pressió urbana o desigualtat en l’ús del territori.

Per dur a terme aquest estudi, ens han proporcionat un conjunt de dades (Country(in).csv) que conté informació detallada sobre la població, la superfície i altres atributs de cada país.


Crea una funció en Python que rebi com a paràmetres:

Un DataFrame amb dades de països.
El nombre n de països que es volen obtenir com a resultat.

La funció ha de calcular, per a cada país, la quantitat de metres quadrats disponibles per persona, utilitzant la següent fórmula:

Metres per persona = SurfaceArea × 1.000.000 / Population

La funció ha de retornar els n països amb menys espai disponible per habitant.

Executa la funció per obtenir els 10 països amb menys espai per persona i després per obtenir-ne 20.

In [326]:
dades = pd.read_csv("csvs/Country(in).csv", sep=";", encoding="latin-1")
def espai(dades, n):
    dades['Metres_persona'] = (dades['SurfaceArea'] * 1000000) / dades['Population']
    return dades.sort_values(by='Metres_persona').head(n)[['Name', 'Metres_persona']]

In [ ]:
#Otra forma de hacerlo con df.nsmallest(top n, columna con elementos a ordenar)--> ordename desde el mas pequeño--> a la inversa nlargest
def espai2(dades, n):
    dades['metres_persona']= (dades['SurfaceArea'] * 1_000_000) / dades['Population']
    resultat= dades.nsmallest(n, 'metres_persona')
    return resultat[['Name', 'metres_persona']]

Name  metres_persona
129                          Macao       38.054968
131                         Monaco       44.117647
92                       Hong Kong      158.507815
186                      Singapore      173.254836
79                       Gibraltar      240.000000
225  Holy See (Vatican City State)      400.000000
28                         Bermuda      815.384615
139                          Malta      831.141504
134                       Maldives     1041.958042
21                      Bangladesh     1114.923929

In [304]:
espai(dades, 10)

Name  Metres_persona
129                          Macao       38.054968
131                         Monaco       44.117647
92                       Hong Kong      158.507815
186                      Singapore      173.254836
79                       Gibraltar      240.000000
225  Holy See (Vatican City State)      400.000000
28                         Bermuda      815.384615
139                          Malta      831.141504
134                       Maldives     1041.958042
21                      Bangladesh     1114.923929

In [302]:
espai(dades, 20)

Name  Metres_persona
129                          Macao       38.054968
131                         Monaco       44.117647
92                       Hong Kong      158.507815
186                      Singapore      173.254836
79                       Gibraltar      240.000000
225  Holy See (Vatican City State)      400.000000
28                         Bermuda      815.384615
139                          Malta      831.141504
134                       Maldives     1041.958042
21                      Bangladesh     1114.923929
23                         Bahrain     1124.797407
31                        Barbados     1592.592593
217                         Taiwan     1625.988497
161                          Nauru     1750.000000
147                      Mauritius     1761.658031
0                            Aruba     1873.786408
176                      Palestine     2017.736214
116                    South Korea     2122.662454
216                         Tuvalu     2166.666667
193                     San Marino     2259.259259

### **Per a treballar més: Funcions**

1. Crea una funció que, donada una frase o una paraula, retorni el nombre de vocals. Un cop definida, sol·licita una paraula o frase a l’usuari i imprimeix el nombre de vocals.

In [385]:
def compte(frase):
    vocals=['a', 'e', 'i', 'o', 'u']
    sumatori=0
    for lletra in frase.lower():
        if lletra in vocals:
            sumatori+=1
    return(sumatori)

recompte=compte(input('Introdueixi una frase o paraula sense accents a continuació: '))
print(f"El nombre de vocals és: {recompte}")

El nombre de vocals és: 2


In [387]:
#Solucion con set, mas completa y eficiente
def compte(frase):
    vocals=set('aeiouàèéíïòóúü') #El set es un conjunto que solo permite elementos unicos. Internamente es como lista, el set descompone caracter por caracter el conjunto.
    sumatori=0
    for lletra in frase.lower():
        if lletra in vocals: #recorre cada caracter del conjunto
            sumatori+=1
    return(sumatori)
recompte=compte(input('Introdueixi una frase o paraula sense accents a continuació: '))
print(f"El nombre de vocals és: {recompte}")

El nombre de vocals és: 3


In [408]:
#Si nos pidieran contar todas las letras...
def funcion(respuesta):
    acumulador=0
    for x in respuesta:
        if x.isalpha(): #detecta letras
            acumulador+=1
    return acumulador
cuenta=funcion(input('Ingresa una palabra o frase: '))
print(f"El nombre de letras és: {cuenta}")

El nombre de letras és: 7


2. Defineix una funció que, donat un número, imprimeixi la seva taula de multiplicar. (Exemple: 1 × 1 = 1, 1 × 2 = 2, 1 × 3 = 3, ...).

In [ ]:
def multi(num):
    tabla=[1,2,3,4,5,6,7,8,9,10]
    resultats=[]
    for numero in tabla:
        operacio=num*numero
        resultats.append(operacio)
    return resultats

num=int(input('Indiqui un número: '))
multiplicacio=multi(num)
print(f'La taula de multiplicar de {num} és: {multiplicacio}')


La taula de multiplicar de 2 és: [2, 4, 6, 8, 10, 12, 14, 16, 18, 20]


In [ ]:
def taula(num):
    for ronda in range(1,11): #for intera en un rango que va del 1 al 10
        operacio=num*ronda
        print(f'{num} x {ronda} = {operacio}')

taula(int(input('Indiqui un número: ')))


6 x 1 = 6
6 x 2 = 12
6 x 3 = 18
6 x 4 = 24
6 x 5 = 30
6 x 6 = 36
6 x 7 = 42
6 x 8 = 48
6 x 9 = 54
6 x 10 = 60


3. Defineix una funció que, donada una frase, retorni el nombre de paraules úniques que conté, sense fer servir la funció len(). (Pots utilitzar el mètode split() d’un string.)

4. Defineix una funció que, donada una llista de números, imprimeixi només els números parells. Genera una llista d’exemple i crida la funció.

5. Escriu una funció que demani un text i retorni un diccionari amb les lletres que apareixen i la quantitat de cops que apareixen, sense fer servir funcions d’agregació com count().

6. Crea una funció que demani 5 noms de plats i el seu nombre de calories. A mesura que s’introdueixen, guarda aquestes dades en un diccionari i retorna’l. Executa la funció i imprimeix el diccionari resultant.

*Apuntes*

Bucles:

for x in df: → columnas

for i, fila in df.iterrows(): → filas

In [ ]:
#Chuletario
def paraula_llarga_i_curta(llistat):
    # Inicializamos variables con la primera palabra
    maxim = llistat[0]
    minim = llistat[0]
    
    for paraula in llistat:
        if len(paraula) > len(maxim):
            maxim = paraula
        if len(paraula) < len(minim):
            minim = paraula
            
    return [maxim, minim]

# Uso genérico
llistat = ['palabra1', 'palabra2', 'palabra3', 'palabra4']
resultat = paraula_llarga_i_curta(llistat)
print("Palabra más larga:", resultat[0])
print("Palabra más corta:", resultat[1])

In [ ]:
def classificar_alumnes(alumnes):
    aprovats = {}
    suspesos = {}
    
    for nom, nota in alumnes.items():
        if nota >= 5:
            aprovats[nom] = nota
        else:
            suspesos[nom] = nota
            
    return aprovats

# Uso genérico con cualquier diccionario
alumnes = {'Alumno1': 7, 'Alumno2': 3, 'Alumno3': 9}
resultat = classificar_alumnes(alumnes)
print("Aprobados:", resultat)

In [ ]:
import pandas as pd

def rendiment(df, columna_vendedor, columna_valor):
    # Agrupar por vendedor y sumar ventas
    totals = df.groupby(columna_vendedor)[columna_valor].sum()
    
    # Calcular la media
    mitjana = totals.mean()
    
    # Lista de vendedores con ventas por debajo de la media
    baix_rendiment = totals[totals < mitjana].index.tolist()
    
    return baix_rendiment

# Uso genérico
# df = pd.read_csv('ruta_al_csv.csv')
# resultat = rendiment(df, 'Vendedor', 'Importe')
# print("Vendedores con rendimiento menor a la media:", resultat)